### ISSUE TITLE
False Positive with Protocol Binding for Type Vars

### ISSUE THREAD

**Issue by max-muoto** [microsoft/pyright#8494]
I ran into a scenario such as as the following, where we have a subclass with a generic param bound to a protocol:

```python
from typing import Protocol
import abc

class Dumpable(Protocol):
    def dump(self) -> str:
        ...

class BaseContainer[D](abc.ABC):
    item: D

class Container[D: Dumpable](BaseContainer[D]):
    def __init__(self, obj: D) -> None:
        self.item = obj


def foo[S](obj: BaseContainer[S]) -> S:
    return obj.item

# Would expect an error for "str" not implementing `dump` here
foo(Container("test")) #  Argument corresponds to parameter "obj" in function "foo" Argument type is "Container[Unknown]"
```

Functions that accept the parent class bound to an argument falsely report that the type parameter for the argument can't be determined when the subclass is passed in. This doesn't change if `D` in `BaseContainer` is also bound to `Dumpable`, for example:

```python
from typing import Protocol
import abc

class Dumpable(Protocol):
    def dump(self) -> str:
        ...

class BaseContainer[D: Dumpable](abc.ABC):
    def __init__(self, obj: D) -> None:
        self.item = obj

class Container[D: Dumpable](BaseContainer[D]):
    def __init__(self, obj: D) -> None:
        self.item = obj


def foo[S: Dumpable](obj: BaseContainer[S]) -> S:
    return obj.item

foo(Container("test"))
```

[Pyright Playground
](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAK4MYAxmADYBQWOeUAhgEalVWkUMDOXUAIgFdszCgFMAFETAlyFAJQAuKlBVQAJqODqhCcV1EVgcqAFoAfFC4wQS1XagA6J2w7deAIW6iAwmBQwGVFEQAG0%2BAF1xZlIHAEF3b0VlVSQYUQgFfhdOHihff0CUYLDMwWEmMUjPfXyAoNCIpLsNLQB9VtRU9r0DYAAaKDAmACtS43MoADk-UVt7FX1DB1T0qABeQZG2KhaoYDAwEIBlSKHRqGqfPzqi0JPxiyO5lRBRGAEQFE3h5bSINjsVH2YHEtUKwXEACI0lZIXJjEA)

Pyright Version: 1.1.372
Python Version: 3.12

**Comment by erictraut** [comment]
Thanks for the bug report.

It turns out this was already fixed by another change I recently made. This bug will be addressed in the next release of pyright.

**Comment by max-muoto** [comment]
> Thanks for the bug report.
> 
> It turns out this was already fixed by another change I recently made. This bug will be addressed in the next release of pyright.

Great, thanks!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.373.
